runtime/bcutil/defineclass.c

```diff
@@ -861,4 +885,111 @@ classCouldPossiblyBeShared(J9VMThread * vmThread, J9LoadROMClassData * loadData)
+		rc = TRUE;
+	} else {
+		const U_8 *anonClassName = J9UTF8_DATA(J9ROMCLASS_CLASSNAME(anonROMClass));
+		const U_8 *hostClassName = J9UTF8_DATA(J9ROMCLASS_CLASSNAME(hostROMClass));
+		const UDATA hostClassPackageNameLength = packageNameLength(hostROMClass);
+		if (J9UTF8_DATA_EQUALS(anonClassName, anonClassPackageNameLength, hostClassName, hostClassPackageNameLength)) {
+			rc = TRUE;
+		}
+	}
+
+	return rc;
+}
+
+/* Create error message with host class and anonymous class. */
+static char*
+createErrorMessage(J9VMThread *vmStruct, J9ROMClass *anonROMClass, J9ROMClass *hostROMClass, const char* errorMsg) {
+	PORT_ACCESS_FROM_VMC(vmStruct);
+	char *buf = NULL;
+
+	if (NULL != errorMsg) {
+		UDATA bufLen = 0;
+		const J9UTF8 *anonClassName = J9ROMCLASS_CLASSNAME(anonROMClass);
+		const J9UTF8 *hostClassName = J9ROMCLASS_CLASSNAME(hostROMClass);
+		const U_8 *hostClassNameData = J9UTF8_DATA(hostClassName);
+		const U_8 *anonClassNameData = J9UTF8_DATA(anonClassName);
+		const UDATA hostClassNameLength = J9UTF8_LENGTH(hostClassName);
+
+		/* Anonymous class name has trailing digits. Example - "test/DummyClass/00000000442F098".
+		 * The code below removes the trailing digits, "/00000000442F098", from the anonymous class name.
+		 */
+		UDATA anonClassNameLength = J9UTF8_LENGTH(anonClassName) - 1;
+		for (; anonClassNameLength >= 0; anonClassNameLength--) {
+			if (anonClassNameData[anonClassNameLength] == '/') {
+				break;
+			}
+		}
+
+		bufLen = j9str_printf(PORTLIB, NULL, 0, errorMsg,
+						hostClassNameLength, hostClassNameData,
+						anonClassNameLength, anonClassNameData);
+		if (bufLen > 0) {
+			buf = j9mem_allocate_memory(bufLen, OMRMEM_CATEGORY_VM);
+			if (NULL != buf) {
+				j9str_printf(PORTLIB, buf, bufLen, errorMsg,
+						hostClassNameLength, hostClassNameData,
+						anonClassNameLength, anonClassNameData);
+			}
+		}
+	}
+
+	return buf;
+}
+
+/* From Java 9 and onwards, set IllegalArgumentException when host class and anonymous class have different packages. */
+static void
+setIllegalArgumentExceptionHostClassAnonClassHaveDifferentPackages(J9VMThread *vmStruct, J9ROMClass *anonROMClass, J9ROMClass *hostROMClass) {
+	PORT_ACCESS_FROM_VMC(vmStruct);
+	const J9JavaVM *vm = vmStruct->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	/* Construct error string */
+	const char *errorMsg = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_HOSTCLASS_ANONCLASS_DIFFERENT_PACKAGES, NULL);
+	char *buf = createErrorMessage(vmStruct, anonROMClass, hostROMClass, errorMsg);
+	vmFuncs->setCurrentExceptionUTF(vmStruct, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, buf);
+	j9mem_free_memory(buf);
+}
+
+/* Free the memory segment corresponding to the anonymous ROM class. */
+static void
+freeAnonROMClass(J9JavaVM *vm, J9ROMClass *romClass) {
+	if (NULL != romClass) {
+		J9MemorySegment **previousSegmentPointerROM = &vm->anonClassLoader->classSegments;
```

Comment:
This should enter the javaVM->classMemorySegments->segmentMutex first

Reply:
added